language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -77,45 +73,117 @@ import           Data.HashMap.Monoidal
 import qualified Data.Map                                           as Map
 import           Data.Text.Prettyprint.Doc
 
+import           Data.Array
+
+builtinNameArities :: Array BuiltinName Int
+builtinNameArities =
+    listArray (minBound, maxBound) $
+        [minBound..maxBound] <&> \name ->
+            withTypedBuiltinName @_ @(Term TyName Name DefaultUni ()) name $
+                \(TypedBuiltinName _ sch) -> countArgs sch
+{-# NOINLINE builtinNameArities #-}  -- Just in case.
+
 {- Note [Scoping]
 The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
 prerequisite. The CEK machine correctly handles name shadowing.
 -}
 
+type TermWithMem uni = WithMemory Term uni
+type TypeWithMem uni = Type TyName uni ExMemory
+type KindWithMem = Kind ExMemory
+
+-- 'Values' for the modified CEK machine.
+data CekVal uni =
+    -- TODO: we probably want to store a @Some (ValueOf uni)@ here, but then we have trouble in
+    -- 'readKnownCek'. I'll reconsider the way we deal with annotations once again.
```

Comment:
Dang, I recall I had an idea, but I can't remember what that idea was anymore.

Reply:
I know the feeling...